README.md
# WSSE

WSSE Username Token generator for Node.js

See: <http://www.xml.com/pub/a/2003/12/17/dive.html>

## Installation

```bash
npm install wsse
```

## Usage

```javascript
import wsse, { UsernameToken } from 'wsse';

const token = wsse({ username: 'bob', password: 'taadtaadpstcsm' });

// 'bob'
console.log(token.getUsername());

// 'taadtaadpstcsm'
console.log(token.getPassword());

// e.g. '2003-12-15T14:43:07Z'
console.log(token.getCreated());

// e.g. 'd36e316282959a9ed4c89851497a717f'
console.log(token.getNonce());

// e.g. 'quR/EWLAV4xLf9Zqyw4pDmfV9OY='
console.log(token.getPasswordDigest());

// e.g. 'UsernameToken Username="bob", PasswordDigest="quR/EWLAV4xLf9Zqyw4pDmfV9OY=", Nonce="d36e316282959a9ed4c89851497a717f", Created="2003-12-15T14:43:07Z"'
console.log(token.getWSSEHeader());
console.log(token.toString());
console.log(token + '');

// ----- advanced -----

// you can use `UsernameToken` class.
const token2 = new UsernameToken({
  username: 'bob',                           // (required)
  password: 'taadtaadpstcsm',                // (required)
  created: '2003-12-15T14:43:07Z',           // (optional) you can specify `craeted`.
  nonce: 'd36e316282959a9ed4c89851497a717f', // (optional) you can specify `nonce`.
  sha1encoding: 'hex'                        // (optional) you can specify `sha1encoding` for wrong WSSE Username Token implementation.
});

// you can use `nonceBase64` option. it encodes the nonce to base64 in header.
// 'UsernameToken Username="bob", PasswordDigest="quR/EWLAV4xLf9Zqyw4pDmfV9OY=", Nonce="ZDM2ZTMxNjI4Mjk1OWE5ZWQ0Yzg5ODUxNDk3YTcxN2Y=", Created="2003-12-15T14:43:07Z"'
// e.g. Hatena API http://developer.hatena.ne.jp/ja/documents/auth/apis/wsse
console.log(token2.getWSSEHeader({ nonceBase64: true }));
console.log(token2.toString({ nonceBase64: true }));
```

## Badges

[![npm version][npm-badge-url]][npm-url]
[![Coveralls][coveralls-badge-url]][coveralls-url]

[coveralls-badge-url]: https://img.shields.io/coveralls/github/bouzuya/node-wsse.svg
[coveralls-url]: https://coveralls.io/github/bouzuya/node-wsse
[npm-badge-url]: https://img.shields.io/npm/v/wsse.svg
[npm-url]: https://www.npmjs.com/package/wsse

## License

### >=2.0.0

[MIT](LICENSE)

### <1.0.0

ISC

## Author

[bouzuya][user] &lt;[m@bouzuya.net][email]&gt; ([https://bouzuya.net][url])

[user]: https://github.com/bouzuya
[email]: mailto:m@bouzuya.net
[url]: https://bouzuya.net


CHANGELOG.md
# [6.0.0](https://github.com/bouzuya/node-wsse/compare/v5.1.0...v6.0.0) (2023-02-18)

- Remove support CommonJS

## [5.1.0](https://github.com/bouzuya/node-wsse/compare/v5.0.0...v5.1.0) (2020-06-30)

- fix: add `import { Buffer } from "buffer";` ([2a978f5](https://github.com/bouzuya/node-wsse/commit/2a978f5))
- build: add rollup settings ([1a1ee27](https://github.com/bouzuya/node-wsse/commit/1a1ee27))

# [5.0.0](https://github.com/bouzuya/node-wsse/compare/v4.0.1...v5.0.0) (2020-04-19)

- build!: drop support for node.js v8 ([6098a6f](https://github.com/bouzuya/node-wsse/commit/6098a6f))
- build(deps-dev): update dev deps
- style: use prettier

### BREAKING CHANGE

- drop support for node.js v8

## [4.0.1](https://github.com/bouzuya/node-wsse/compare/v4.0.0...v4.0.1) (2019-12-12)

# [4.0.0](https://github.com/bouzuya/node-wsse/compare/v3.0.1...v4.0.0) (2019-06-06)

## [3.0.1](https://github.com/bouzuya/node-wsse/compare/v3.0.0...v3.0.1) (2019-06-06)

# [3.0.0](https://github.com/bouzuya/node-wsse/compare/2.4.0...v3.0.0) (2019-01-22)

# [2.4.0](https://github.com/bouzuya/node-wsse/compare/2.3.0...2.4.0) (2016-11-10)

# [2.3.0](https://github.com/bouzuya/node-wsse/compare/2.2.0...2.3.0) (2015-12-09)

# [2.2.0](https://github.com/bouzuya/node-wsse/compare/2.1.0...2.2.0) (2015-12-09)

# [2.1.0](https://github.com/bouzuya/node-wsse/compare/2.0.0...2.1.0) (2015-06-13)

# [2.0.0](https://github.com/bouzuya/node-wsse/compare/1.0.0...2.0.0) (2015-06-12)

# [1.0.0](https://github.com/bouzuya/node-wsse/compare/0.0.3...1.0.0) (2015-06-11)


test/wsse.js
var expect = require('chai').expect;
var wsse = require('../lib/wsse')();

// http://www.xml.com/pub/a/2003/12/17/dive.html

describe('Wsse', function() {

  describe('getNonce', function() {

    it('works', function(done) {
      var nonce = wsse.getNonce();
      expect(nonce).to.have.property('created');
      expect(nonce).to.have.property('nonce');
      done();
    });

  });

  describe('getPasswordDigest', function() {

    it('works', function(done) {
      var expected = 'quR/EWLAV4xLf9Zqyw4pDmfV9OY=';
      var password = 'taadtaadpstcsm';
      var nonce = 'd36e316282959a9ed4c89851497a717f';
      var created = '2003-12-15T14:43:07Z';
      var actual = wsse.getPasswordDigest(nonce, created, password);
      expect(actual).to.equal(expected);
      done();
    });

  });

  describe('getUsernameToken', function() {

    it('works', function() {
      var original = wsse.getNonce;
      wsse.getNonce = function() {
        return { created: '2003-12-15T14:43:07Z', nonce: 'd36e316282959a9ed4c89851497a717f' };
      };

      var expected = 'Username="bob", PasswordDigest="quR/EWLAV4xLf9Zqyw4pDmfV9OY=", Nonce="d36e316282959a9ed4c89851497a717f", Created="2003-12-15T14:43:07Z"';
      var username = 'bob';
      var password = 'taadtaadpstcsm';
      var actual = wsse.getUsernameToken(username, password);
      expect(actual).to.equal(expected);

      wsse.getNonce = original;
    });

  });

});


test/index.ts
import { tests as emarsysTests } from "./emarsys.js";
import { Test, run } from "./helper.js";
import { tests as standardTests } from "./standard.js";

const tests = ([] as Test[]).concat(emarsysTests).concat(standardTests);

run(tests).catch(() => process.exit(1));


test/helper.ts
import { Test, run } from "beater";
import { fixture } from "beater-helpers/fixture.js";

const test = (name: string, fn: Test): Test => {
  Object.defineProperty(fn, "name", { value: name });
  return fn;
};

export { Test, fixture, run, test };


src/index.ts
import { createHash, randomFillSync } from "node:crypto";
import { Buffer } from "node:buffer";

const ensure = <T>(v: T | null | undefined, d: T): T => {
  return typeof v !== "undefined" && v !== null ? v : d;
};

type SHA1Encoding = "base64" | "latin1" | "hex";

class UsernameToken {
  private _created: string;
  private _nonce: string;
  private _password: string;
  private _sha1encoding: SHA1Encoding | undefined;
  private _username: string;

  constructor({
    username,
    password,
    created,
    nonce,
    sha1encoding,
  }: {
    username: string;
    password: string;
    created?: string;
    nonce?: string;
    // YOU SHOULD NOT USE THIS OPTION.
    // See: https://github.com/bouzuya/node-wsse/issues/1
    sha1encoding?: SHA1Encoding;
  }) {
    this._username = username; // required
    this._password = password; // required
    this._created = ensure(created, this._newCreated());
    this._nonce = ensure(nonce, this._newNonce());
    this._sha1encoding = sha1encoding;
  }

  public getCreated(): string {
    return this._created;
  }

  public getNonce(): string {
    return this._nonce;
  }

  public getNonceBase64(): string {
    return this._base64ForString(this._nonce);
  }

  public getPassword(): string {
    return this._password;
  }

  public getPasswordDigest(): string {
    const text = this.getNonce() + this.getCreated() + this.getPassword();
    const buf = this._sha1(text);
    return typeof this._sha1encoding === "undefined"
      ? this._base64(buf)
      : // workaround for emarsys bad implementation
        this._base64ForString(buf.toString(this._sha1encoding));
  }

  public getUsername(): string {
    return this._username;
  }

  public getWSSEHeader(options?: { nonceBase64?: boolean }): string {
    const opts = ensure<{ nonceBase64?: boolean }>(options, {});
    // workaround for hatena api
    const nonceBase64 = ensure<boolean>(opts.nonceBase64, false);
    return (
      "UsernameToken " +
      [
        `Username="${this.getUsername()}"`,
        `PasswordDigest="${this.getPasswordDigest()}"`,
        `Nonce="${nonceBase64 ? this.getNonceBase64() : this.getNonce()}"`,
        `Created="${this.getCreated()}"`,
      ].join(", ")
    );
  }

  public newToken(): UsernameToken {
    // new nonce & new created
    return new UsernameToken({
      password: this._password,
      sha1encoding: this._sha1encoding,
      username: this._username,
    });
  }

  public toString(options?: { nonceBase64?: boolean }): string {
    return this.getWSSEHeader(options);
  }

  private _base64(b: Buffer): string {
    return b.toString("base64");
  }

  private _base64ForString(s: string): string {
    return this._base64(Buffer.from(s, "utf-8"));
  }

  private _newCreated(): string {
    return new Date().toISOString();
  }

  private _newNonce(): string {
    const buf = Buffer.alloc(10);
    randomFillSync(buf);
    return buf.toString("hex");
  }

  private _sha1(s: string): Buffer {
    const sha1 = createHash("sha1");
    sha1.update(s, "utf-8");
    return sha1.digest();
  }
}

const usernameToken = ({
  username,
  password,
}: {
  username: string;
  password: string;
}): UsernameToken => {
  return new UsernameToken({ username, password });
};

export default usernameToken;
export { UsernameToken };


lib/username-token.js
var UsernameToken, crypto;

crypto = require('crypto');

UsernameToken = (function() {
  function UsernameToken(arg) {
    var created, nonce, password, ref, sha1encoding, username;
    ref = arg != null ? arg : {}, username = ref.username, password = ref.password, created = ref.created, nonce = ref.nonce, sha1encoding = ref.sha1encoding;
    this._username = username;
    this._password = password;
    this._created = created != null ? created : this._newCreated();
    this._nonce = nonce != null ? nonce : this._newNonce();
    this._sha1encoding = sha1encoding != null ? sha1encoding : void 0;
  }

  UsernameToken.prototype.getCreated = function() {
    return this._created;
  };

  UsernameToken.prototype.getNonce = function() {
    return this._nonce;
  };

  UsernameToken.prototype.getNonceBase64 = function() {
    return this._base64(this._nonce);
  };

  UsernameToken.prototype.getPassword = function() {
    return this._password;
  };

  UsernameToken.prototype.getPasswordDigest = function() {
    var text;
    text = this.getNonce() + this.getCreated() + this.getPassword();
    return this._base64(this._sha1(text, this._sha1encoding));
  };

  UsernameToken.prototype.getUsername = function() {
    return this._username;
  };

  UsernameToken.prototype.getWSSEHeader = function(arg) {
    var nonceBase64;
    nonceBase64 = (arg != null ? arg : {}).nonceBase64;
    nonceBase64 = nonceBase64 != null ? nonceBase64 : false;
    return 'UsernameToken ' + ["Username=\"" + (this.getUsername()) + "\"", "PasswordDigest=\"" + (this.getPasswordDigest()) + "\"", "Nonce=\"" + (nonceBase64 ? this.getNonceBase64() : this.getNonce()) + "\"", "Created=\"" + (this.getCreated()) + "\""].join(', ');
  };

  UsernameToken.prototype.newToken = function() {
    return new UsernameToken({
      username: this._username,
      password: this._password,
      sha1encoding: this._sha1encoding
    });
  };

  UsernameToken.prototype.toString = function(arg) {
    var nonceBase64;
    nonceBase64 = (arg != null ? arg : {}).nonceBase64;
    return this.getWSSEHeader({
      nonceBase64: nonceBase64
    });
  };

  UsernameToken.prototype._base64 = function(s) {
    return new Buffer(s).toString('base64');
  };

  UsernameToken.prototype._sha1 = function(s, encoding) {
    var sha1;
    sha1 = crypto.createHash('sha1');
    sha1.update(s, 'utf-8');
    return sha1.digest(encoding);
  };

  UsernameToken.prototype._newCreated = function() {
    return new Date().toISOString();
  };

  UsernameToken.prototype._newNonce = function() {
    return [0, 1, 2, 3, 4, 5, 6, 7, 8, 9].map(function() {
      return Math.floor(Math.random() * 256).toString(16);
    }).join('');
  };

  return UsernameToken;

})();

module.exports = function(arg) {
  var password, ref, username;
  ref = arg != null ? arg : {}, username = ref.username, password = ref.password;
  return new UsernameToken({
    username: username,
    password: password
  });
};

module.exports.UsernameToken = UsernameToken;


lib/wsse.js
var crypto = require('crypto');

var Wsse = function() {};

Wsse.prototype.getNonce = function() {
  var created = new Date().toISOString();
  var nonce = '';
  for(var i = 0; i < 10; i++) {
    var r = Math.floor(Math.random() * 256);
    nonce += r.toString(16);
  }
  return { created: created, nonce: nonce };
};

Wsse.prototype.getPasswordDigest = function(nonce, created, password) {
  var input = nonce + created + password;
  var sha1 = crypto.createHash('sha1');
  sha1.update(input, 'utf-8');
  return sha1.digest('base64');
};

Wsse.prototype.base64 = function(s) {
  return new Buffer(s).toString('base64');
};

Wsse.prototype.getUsernameToken = function(username, password, options) {
  options = options || {};
  var nonceBase64 = !!options.nonceBase64 || false;
  var nonceObj = this.getNonce();
  var nonce = nonceObj.nonce;
  var created = nonceObj.created;
  var passwordDigest = this.getPasswordDigest(nonce, created, password);
  var headerNonce = nonceBase64 ? this.base64(nonce) : nonce;
  var usernameToken = '';
  usernameToken += 'Username="' + username + '", ';
  usernameToken += 'PasswordDigest="' + passwordDigest + '", ';
  usernameToken += 'Nonce="' + headerNonce + '", ';
  usernameToken += 'Created="' + created + '"';
  return usernameToken;
};

module.exports = function() { return new Wsse(); };

